scripts/globals/job_utils/ranger.lua

```diff
@@ -0,0 +1,277 @@
+
+    return 0, 0
+end
+
+-----------------------------------
+-- Ability Use Functions
+-----------------------------------
+
+xi.job_utils.ranger.useEagleEyeShot = function(player, target, ability)
+ if player:getWeaponSkillType(xi.slot.RANGED) == xi.skill.MARKSMANSHIP then
+        action:setAnimation(target:getID(), action:getAnimation(target:getID()) + 1)
+    end
+
+    local params = {}
+
+    params.numHits = 1
+
+    -- TP params.
+    local tp          = 1000 -- to ensure ftp multiplier is applied
+    params.ftpMod     = { 5.0, 5.0, 5.0 }
+    params.critVaries = { 0.0, 0.0, 0.0 }
+
+    -- Stat params.
+    params.str_wsc = 0
+    params.dex_wsc = 0
+    params.vit_wsc = 0
+    params.agi_wsc = 0
+    params.int_wsc = 0
+    params.mnd_wsc = 0
+    params.chr_wsc = 0
+
+    params.enmityMult = 0.5
+
+    -- Job Point Bonus Damage
+    local jpValue = player:getJobPointLevel(xi.jp.EAGLE_EYE_SHOT_EFFECT)
+    player:addMod(xi.mod.ALL_WSDMG_ALL_HITS, jpValue * 3)
+
+    local damage, _, tpHits, extraHits = xi.weaponskills.doRangedWeaponskill(player, target, 0, params, tp, action, true)
+
+    -- Set the message id ourselves
+    if tpHits + extraHits > 0 then
+        action:messageID(target:getID(), xi.msg.basic.JA_DAMAGE)
+        action:speceffect(target:getID(), 32)
+    else
+        action:messageID(target:getID(), xi.msg.basic.JA_MISS_2)
+        action:speceffect(target:getID(), 0)
+    end
+
+    return damage
+end
+
+xi.job_utils.ranger.useVelocityShot = function(player, target, ability)
+	player:addStatusEffect(xi.effect.VELOCITY_SHOT, 1, 0, 7200)
+end
+
+xi.job_utils.ranger.useSharpshot = function(player, target, ability)
+    local power = 40 + player:getMod(xi.mod.SHARPSHOT)
+    player:addStatusEffect(xi.effect.SHARPSHOT, power, 0, 60)
+end
+
+xi.job_utils.ranger.useScavenge = function(player, target, ability)
+    -- RNG AF2 quest check
+    local fireAndBrimstoneCS = player:getCharVar('fireAndBrimstone')
+
+    if
+        player:getZoneID() == xi.zone.CASTLE_OZTROJA and fireAndBrimstoneCS == 5 and-- zone + quest match
+        not player:hasItem(xi.item.OLD_EARRING) and -- make sure player doesn't already have the earring
+        player:getYPos() > -43 and player:getYPos() < -38 and -- Y match
+        player:getXPos() > -85 and player:getXPos() < -73 and -- X match
+        player:getZPos() > -85 and player:getZPos() < -75 and -- Z match
+        math.random(1, 100) <= 50
+    then
+		npcUtil.giveItem(player, xi.item.OLD_EARRING)
+
+    else
+        local bonuses        = (player:getMod(xi.mod.SCAVENGE_EFFECT) + player:getMerit(xi.merit.SCAVENGE_EFFECT)) / 100
+        local arrowsToReturn = math.floor(math.floor(player:getLocalVar('ArrowsUsed') % 10000) * (player:getMainLvl() / 200 + bonuses))
+        local playerID       = target:getID()
+
+        if arrowsToReturn == 0 then
+            action:messageID(playerID, 139)
```

Comment:
Same: Did you test this, action doesn't exist in this scope?

Reply:
Will fix same as eagle eye shot response, my mistake and not being careful enough.